kernel/syscall.c

```diff
@@ -200,12 +200,63 @@ static const char* _format_timespec(
     return buf->data;
 }
 
+static bool _trace_syscall(long n)
+{
+    // Check if syscall tracing is enabled.
+    if (__myst_kernel_args.strace_config.trace_syscalls)
+    {
+        if (__myst_kernel_args.strace_config.filter)
+        {
+            // If filtering is enabled, trace only this syscall has been
+            // specified in the filter.
+            return __myst_kernel_args.strace_config.trace[n];
+        }
+        else
+        {
+            // Trace all syscalls.
+            return true;
+        }
+    }
+    return false;
+}
+
+static void _syscall_failure_hook(long n, long ret)
+{
+    // Set breakpoint in this function to stop execution when a syscall fails.
+    // Set condition for the breakpoint to control which syscall failures
+    // trigger the breakpoint.
+    (void)n;
+    (void)ret;
+}
+
+static bool _trace_syscall_return(long n, long ret)
+{
+    // If this syscall has been configured to be traced, then trace the return
+    // too.
+    if (_trace_syscall(n))
+        return true;
+
+    // Check if tracing failing syscalls has been enabled.
+    if (__myst_kernel_args.strace_config.trace_failing)
+    {
+        // If the syscall returns a negative value and has a valid error name,
+        // then consider it a failure and enable tracing.
+        if (ret < 0)
+        {
+            const char* error_name = myst_error_name(-ret);
```

Comment:
I guess I should have called this function something closer to the standard C name like ``myst_strerror()``.